File: composition/Scala/composition-pipeline/pipeline/src/main/scala/DomainModel.scala
import common.Pipeline.Result
import common.StringExtensions._
import shared.StringExtensions._

// INFO: An example of a domain model for an simple application. The idea is to provide sample methods that we can chain
//       together using the pipeline pattern.
object DomainModel {
  val INVALID_NAME_MESSAGE = "Invalid name"
  val EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE = "Email address format is incorrect"

  case class Person(name: String = "", emailAddress: String = "")

  def validateName(person: Person): Result[Person] = {
    person.name match {
      case name if name.isNullOrEmpty => Result(INVALID_NAME_MESSAGE, person)
      case _ => Result(data = person)
    }
  }

  def validateEmailAddress(person: Person): Result[Person] = {
    person.emailAddress match {
      case emailAddress if !emailAddress.isValidEmailAddress => Result(EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE, person)
      case _ => Result(data = person)
    }
  }

  def updateName(name: String) = {
    val apply = (data: Person) => {
      Result(data = Person(name, data.emailAddress))
    }: Result[Person]
    apply
  }

  def updateEmailAddress(emailAddress: String) = {
    val apply = (data: Person) => {
      Result(data = Person(data.name, emailAddress))
    }: Result[Person]
    apply
  }
}

File: composition/C#/CompositionPipeline/DomainModel.cs
using static CompositionPipeline.Shared.StringExtensions;
using CompositionPipeline.Common;

namespace CompositionPipeline
{
	public static class DomainModel
	{
		public const string INVALID_NAME_MESSAGE = "Invalid name";
		public const string EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE = "Email address format is incorrect";

		public static Result<Person> ValidateName(Person person) {
			if (string.IsNullOrWhiteSpace(person.Name))
				return new Result<Person>(INVALID_NAME_MESSAGE, person);

			return new Result<Person>(data: person);
		}

		public static Result<Person> ValidateEmailAddress(Person person) {
			if (person.EmailAddress.IsValidEmailAddress())
				return new Result<Person>(EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE, person);
			return new Result<Person>(data: person);
		}
	}
}
